### PR TITLE
refactor(headermenu): added typescript types to headermenu

### LINKED ISSUE #13589: Add TypeScript types to HeaderMenu

### PR DESCRIPTION
**Pull request by Gururajj77**
Closes #13589 

Added typescript types for HeaderMenu component

#### Changelog

**New**

- Added types to the props

**Changed**

- file name from js to tsx

#### Testing / Reviewing

Opened new PR as #16104  had commits which were unrelated.

No new observations from testing perspective.

### CODE REVIEW

File: packages/react/src/components/UIShell/HeaderMenu.tsx [hunk 1]
```diff
@@ -21,7 +21,29 @@ import { composeEventHandlers } from '../../tools/events';
...
+  focusRef?: React.Ref<any>;
+  isActive?: boolean;
+  isCurrentPage?: boolean;
+  menuLinkName: string;
+  onBlur?: (event: React.FocusEvent<HTMLLIElement>) => void;
+  onClick?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
+  onKeyDown?: (event: React.KeyboardEvent<HTMLLIElement>) => void;
+  renderMenuContent?: () => JSX.Element;
+  tabIndex?: number;
+  children?: React.ReactNode;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/HeaderMenu.tsx]: Copying the jsdoc comments from proptypes to be up here as well would be ideal. They're [important for intellisense](https://code.visualstudio.com/docs/typescript/typescript-editing#_jsdoc-support)

File: packages/react/src/components/UIShell/HeaderMenu.tsx [hunk 2]
```diff
@@ -228,9 +257,9 @@ class HeaderMenu extends React.Component {
     };
     const itemClassName = cx({
       [`${prefix}--header__submenu`]: true,
-      [customClassName]: !!customClassName,
+      [customClassName || 'default-class-name']: !!customClassName,
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/HeaderMenu.tsx]:
```suggestion
      [customClassName]: !!customClassName,
```

File: packages/react/src/components/UIShell/HeaderMenu.tsx [hunk 3]
```diff
@@ -286,17 +315,17 @@ class HeaderMenu extends React.Component {
...
+      return React.cloneElement(item as React.ReactElement<any>, {
         ref: this.handleItemRef(index),
       });
     }
   };
 }
 
 const HeaderMenuForwardRef = React.forwardRef((props, ref) => {
-  return <HeaderMenu {...props} focusRef={ref} />;
+  return <HeaderMenu menuLinkName="link" {...props} focusRef={ref} />;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/HeaderMenu.tsx:379]: What's the ts error related to this change?

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Looks pretty good! Just a couple things

File: packages/react/src/components/UIShell/HeaderMenu.tsx (hunk 1 above)
Comment (Gururajj77) [reply packages/react/src/components/UIShell/HeaderMenu.tsx]: will add them from now

File: packages/react/src/components/UIShell/HeaderMenu.tsx (hunk 2 above)
Comment (Gururajj77) [reply packages/react/src/components/UIShell/HeaderMenu.tsx]: A computed property name must be of type 'string', 'number', 'symbol', or 'any'.ts(2464)

we get this error when we remove the default class name. 

I found another way to resolve this, by enclosing the customClassName in string literal.

File: packages/react/src/components/UIShell/HeaderMenu.tsx (hunk 3 above)
Comment (Gururajj77) [reply packages/react/src/components/UIShell/HeaderMenu.tsx:379]: Property 'menuLinkName' is missing in type '{ focusRef: ForwardedRef<unknown>; }' but required in type 'Pick<Readonly<HeaderMenuProps>, keyof HeaderMenuProps>'.

This is the error.

**Review by guidari — APPROVED** [review]
LGTM! 🚀 
Working as expected after tests

**Review by tay1orjones — APPROVED** [review]

**Pull request merged**